### PR TITLE
feat: add Unicode library

### PR DESCRIPTION
**Pull request by alex-chew**
This PR adds basic definitions/functions/lemmas for Unicode values and strings as defined in the Unicode Standard 14.0.0: https://www.unicode.org/versions/Unicode14.0.0/.

(Most/all of this PR is implementation of Chapter 3: https://www.unicode.org/versions/Unicode14.0.0/ch03.pdf)

By submitting this pull request, I confirm that my contribution is made under the terms of the MIT license.

### CODE REVIEW

File: src/Unicode/Utf8EncodingScheme.dfy [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ * sequence itself.
+ *
+ * Because the UTF-8 encoding form already deals in ordered byte sequences, the UTF-8 encoding scheme is trivial. The
+ * byte ordering is already obvious and completely defined by the UTF-8 code unit sequence itself. The UTF-8
+ * encoding scheme is defined merely for completeness of the Unicode character encoding model.
+ *
+ * (Section 3.10 D95)
+ *
+ * TODO: this should refine an abstract UnicodeEncodingScheme module,
+ * but defining such a module seems impossible so long as dafny-lang/dafny#1639 is unresolved.
```
Comment (robin-aws) [inline src/Unicode/Utf8EncodingScheme.dfy:25]: Why does https://github.com/dafny-lang/dafny/issues/1639 block this though? Do you have to have a `calc` statement in the abstract module?

File: src/Unicode/Unicode.dfy [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  predicate {:opaque} IsInAssignedPlane(i: CodePoint) {
+    var plane := (i >> 16) as bv8;
+    plane in ASSIGNED_PLANES
+  }
+
+  // These are actually supersets of the Unicode planes,
+  // since not all code points in a plane are assigned.
+  //
+  // TODO: correctly check code points against the list of assigned code points
+  // (https://www.unicode.org/Public/UCD/latest/ucd/UnicodeData.txt)
```
Comment (robin-aws) [inline src/Unicode/Unicode.dfy:64]: Proving a `forall` would indeed be great, but at a minimum how about a test lemma with sample values from each plane?

File: src/Unicode/Utf8EncodingScheme.dfy [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+include "Unicode.dfy"
+include "Utf8EncodingForm.dfy"
+
+/**
+ * The Unicode encoding scheme that serializes a UTF-8 code unit sequence in exactly the same order as the code unit
+ * sequence itself.
+ *
+ * Because the UTF-8 encoding form already deals in ordered byte sequences, the UTF-8 encoding scheme is trivial. The
+ * byte ordering is already obvious and completely defined by the UTF-8 code unit sequence itself. The UTF-8
+ * encoding scheme is defined merely for completeness of the Unicode character encoding model.
```
Comment (robin-aws) [inline src/Unicode/Utf8EncodingScheme.dfy:20]: And also because `unit8` and `bv8` are not the same type? :)

**Review by robin-aws — CHANGES_REQUESTED** [review]
Really love how readable this all is, and how directly it confirms to the terminology in the standard. I made a couple comments about runtime performance, but they aren't blocking - it's extremely useful even just to have a verified reference implementation like this. Really nice work!

File: src/Unicode/Utf16EncodingForm.dfy [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    var w := ((firstWord & 0x3C0) >> 6) as bv24;
+    var u := (w + 1) as bv24;
+    var v := (u << 16) | (x1 << 10) | x2 as Unicode.ScalarValue;
+    assert {:split_here} true;
+    assert EncodeScalarValueDoubleWord(v) == m;
+    v
+  }
+
+  //
+  // Test lemmas
```
[the remark this answers is not in the document]
Comment (alex-chew) [reply src/Unicode/Utf16EncodingForm.dfy]: Good call, I've moved all the "tests" into the examples directory.

File: src/Unicode/Utf8EncodingScheme.dfy (hunk 1 above)
Comment (alex-chew) [reply src/Unicode/Utf8EncodingScheme.dfy]: It's not a strict blocker, but I've left a more detailed explanation in the TODO in the latest commits.

File: src/Unicode/Unicode.dfy (hunk 2 above)
Comment (alex-chew) [reply src/Unicode/Unicode.dfy]: Good call, done in latest commits.

File: src/Unicode/Utf8EncodingScheme.dfy (hunk 3 above)
Comment (alex-chew) [reply src/Unicode/Utf8EncodingScheme.dfy]: Noted in latest commits.

**Review by robin-aws — APPROVED** [review]
This is totally $ £ €💰

**Review by cpitclaudel — APPROVED** [review]

**Pull request merged**